src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -252,21 +253,22 @@ public static WatcherOperation CombineRange(ImmutableSegmentedList<WatcherOperat
                 {
                     case Kind.WatchFiles:
                         for (var i = 0; i < op._paths.Count; i++)
-                        {
                             fileNamesBuilder.Add(op._paths[i]);
-                        }
 
                         for (var i = 0; i < op._tokens.Count; i++)
-                        {
                             tokensBuilder.Add(op._tokens[i]);
-                        }
                         break;
 
                     case Kind.UnwatchFiles:
                         for (var i = 0; i < op._tokens.Count; i++)
-                        {
                             tokensBuilder.Add(op._tokens[i]);
-                        }
+                        break;
+
+                    case Kind.WatchDirectory:
+                        for (var i = 0; i < op._filters.Length; i++)
+                            filtersBuilder.Add(op._filters[i]);
+
+                        cookiesBuilder.AddRange(op._cookies);
```

Comment:
So I think is going to break things -- the list of cookies in the WatchDirectory case isn't just the list that can get copied around, but it's actually the list that gets filled back in. Put another way, the `List<uint>` object is held by the Context object, and is expected that exact instance gets passed around to the ApplyAsync method so it can be filled in later. That way that same list object gets passed to the UnwatchDirectories later.

Reply:
We really should add some tests here. Specifically, have a test that merging two directories and then unsubscribing them successfully results in no watches once we're done. I expect that to be broken.

@svick's test in https://github.com/dotnet/roslyn/issues/74716 is a pretty good start for the general pattern. Rather than firing events and seeing what's fired, it might be easier to just assert the number of watches on the underlying mocked file change service. (Although making sure events actually fire right wouldn't hurt either.)

Reply:
Backed out the watch directory merging, wasn't going to be easy to solve and found a different way to get some watcher reductions.